### PR TITLE
Re-introduce WM_CLOSE listener, new quit protocol posts a second message to next handler

### PR DESCRIPTION
**Pull request by yaakovschectman**
With this approach, when the last window of an app is closed, the engine sends a message to the framework to check if it is allowed to exit. If it may, the windows lifecycle manager synthesizes a second `WM_CLOSE` message that it will ignore, and so the next registered top level window proc delegate, if any, will be able to process the message. If none do so, the message will be handled by the default window proc, so the app will be able to close.

I was not able to get a full system tray example application running to test this, but I could get an application that stays open when its window is closed and can be seen as a system tray icon as long as it is running, albeit the icon was non-functional. As this repro app still exhibited this behavior when using this engine build, I am reasonably confident in concluding that applications that want to be able to run headless when their windows close will function properly.

Addresses https://github.com/flutter/flutter/issues/123654.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/windows_lifecycle_manager.h
Comment (gspencergoog) [inline shell/platform/windows/windows_lifecycle_manager.h:35]: Maybe `sent_close_messages_`?

File: shell/platform/windows/platform_handler.cc
Comment (gspencergoog) [inline shell/platform/windows/platform_handler.cc:210]:
```suggestion
static ExitType StringToAppExitType(const std::string& string) {
```

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Review by cbracken — APPROVED** [review]
Thanks for reworking this!

**Comment by loic-sharma** [comment]
@yaakovschectman Should we ask our partner team if they'd like us to wait until after they ship to merge this? My understanding is that they're shipping their app soon.

File: shell/platform/windows/windows_lifecycle_manager.h
Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.h:26]: Nit: `HWND` is generally the first parameter for win32 APIs, consider swapping

Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.h:26]: Could you add a quick comment on `Quit` and `WindowProc`?

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.cc:26]: We should maintain the `wparam` and `lparam` values when we redispatch (you can do evil tricks and shove state in these unused values). Here's how the keyboard does it:

https://github.com/flutter/engine/blob/252f6cfd0335918d4147ba63e52b5800c7254478/shell/platform/windows/keyboard_manager.cc#L118-L120

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.cc:25]: They keyboard logs an error if there are too many pending redispatches messages as that indicates redispatching is broken. Should we do that as well?

File: shell/platform/windows/platform_handler.h
Comment (loic-sharma) [inline shell/platform/windows/platform_handler.h:49]: Consider updating the comment to mention that `HWND` is the window that initiated the app exit.

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.cc:37]: Please add a comment here about why this is necessary. (Framework shutdown is async, we consume the initial `WM_CLOSE` and redispatch it, so need to ignore redispatched `WM_CLOSE`s).

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.cc:49]: Is this right? Should we leave the `WM_CLOSE` message unhandled if this isn't the last window of the process?

File: shell/platform/windows/platform_handler.cc
Comment (loic-sharma) [inline shell/platform/windows/platform_handler.cc:375]: Oh wow, I didn't realize `HWND` was optional until very late in the review. This is not at all obvious and will result in crashes if folks miss this. Please make all these `HWND`s `std::optional<HWND>`s

File: shell/platform/windows/flutter_windows_engine_unittests.cc
Comment (loic-sharma) [inline shell/platform/windows/flutter_windows_engine_unittests.cc:653]: Can we check the message's content?

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (yaakovschectman) [reply shell/platform/windows/windows_lifecycle_manager.cc:25]: For the keyboard, it should be impossible to get consecutive keyup msgs without a keydown in between, right? But I don't think there's any such guarantee for windows, e.g. if the close button is clicked the same instant as alt+f4 is pressed, so I am not so sure this should be an error state.

File: shell/platform/windows/flutter_windows_engine_unittests.cc
Comment (yaakovschectman) [reply shell/platform/windows/flutter_windows_engine_unittests.cc:653]: We can, albeit perhaps a bit clunkily. What do you think we should check it for?

File: shell/platform/windows/flutter_windows_engine_unittests.cc
Comment (loic-sharma) [inline shell/platform/windows/flutter_windows_engine_unittests.cc:734]: `when it returns false` is a bit unclear, consider something like `when the lifecycle manager doesn't handle a message`

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (loic-sharma) [reply shell/platform/windows/windows_lifecycle_manager.cc:25]: The error is to catch a different scenario: we're redispatching messages but not receiving them. An unbalance would result us in dropping `WM_CLOSE` messages unnecessarily.

File: shell/platform/windows/flutter_windows_engine_unittests.cc
Comment (loic-sharma) [reply shell/platform/windows/flutter_windows_engine_unittests.cc:653]: The concern would be that we accidentally regress the messages' contents.

Why would it be clunky? Can you not cast it to a `std::string` and then use `EXPECT_EQ`? For example:

https://github.com/flutter/engine/blob/252f6cfd0335918d4147ba63e52b5800c7254478/shell/platform/windows/platform_handler_unittests.cc#L113

https://github.com/flutter/engine/blob/252f6cfd0335918d4147ba63e52b5800c7254478/shell/platform/windows/platform_handler_unittests.cc#L200-L202

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (yaakovschectman) [reply shell/platform/windows/windows_lifecycle_manager.cc:25]: What would be the condition that you are concerned may arise in this case, that the value held by the map exceeds 1? If so, could we not need to redispatch 2 messages at a time?

File: shell/platform/windows/flutter_windows_engine_unittests.cc
Comment (yaakovschectman) [reply shell/platform/windows/flutter_windows_engine_unittests.cc:653]: Yes, I was just referring to expecting the json message object to be equal to the string. I think we can know the order, etc., ahead of time, but it might not be obvious, since multiple valid strings would represent the correct message.

Comment (yaakovschectman) [reply shell/platform/windows/flutter_windows_engine_unittests.cc:655]: `The plain [HashMap](https://api.dart.dev/be/180791/dart-collection/HashMap-class.html) is unordered (no order is guaranteed),`
This leads be to believe that the `jsonString` created at https://github.com/flutter/engine/pull/40802/files/ec79b21b9a93211115228da555d0d02a5c227287#diff-0ec6bbe0e6c9b89b68745c51c5e6beda1692354554e97b6c1789bb4614ae8fdfR103 and below does not have a guaranteed order, so testing for equality with a string literal would not be as simple as testing the list example you provided, which is ordered.

Comment (loic-sharma) [reply shell/platform/windows/flutter_windows_engine_unittests.cc:655]: Hm that's a fair concern. We have many tests that assume the order is stable. If Dart ever changes how hash map's order is defined, our tests that use string equality checks on JSON will break and we'll need to fix them.

Regardless, we should verify the message is expected. Feel free to choose whichever approach you prefer :)

Comment (cbracken) [reply shell/platform/windows/flutter_windows_engine_unittests.cc:664]: Dart's default Map class is not a HashMap but a LinkedHashMap, which guarantees order.

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (loic-sharma) [reply shell/platform/windows/windows_lifecycle_manager.cc]: If we redispatch `WM_CLOSE` messages we expect they'll be handled shortly thereafter. It would be unexpected for us to have many redispatched `WM_CLOSE` messages in flight. This may indicate that our redispatch logic is broken. However, I don't feel strongly about asserting for this scenario.

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.cc:48]: Consider:

```suggestion
    // send a request to the framework to see if the app should exit. If it
```

File: shell/platform/windows/windows_lifecycle_manager.h
Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.h:34]: What do you think of having two separate APIs here?

1. Close window
2. Quit app

Apologies for not suggesting this earlier, didn't think of this until now :)

File: shell/platform/windows/flutter_windows_engine.h
Comment (loic-sharma) [inline shell/platform/windows/flutter_windows_engine.h:267]: This API is only called when we receive a `WM_CLOSE` right? If so, we will always have `hwnd`, `wparam`, and `lparam` values and these don't need to be `std::optional`. Please let me know if I'm missing something!

File: shell/platform/windows/windows_lifecycle_manager.cc
Comment (loic-sharma) [inline shell/platform/windows/windows_lifecycle_manager.cc:29]: We should only ignore `WM_CLOSE` messages whose `wparam` and `lparam` matches what we posted.

File: shell/platform/windows/flutter_windows_engine.h
Comment (yaakovschectman) [reply shell/platform/windows/flutter_windows_engine.h:267]: This is also called when `exitApplication` is called on the app side

File: shell/platform/windows/windows_lifecycle_manager.h
Comment (yaakovschectman) [reply shell/platform/windows/windows_lifecycle_manager.h:34]: We currently intend for the two to both be controlled by the `didRequestAppExit` framework-side callback, which is why they are handled by the same engine-side method as well.

File: shell/platform/windows/flutter_windows_engine.h
Comment (loic-sharma) [reply shell/platform/windows/flutter_windows_engine.h:267]: Gotcha. Please update the comment as it currently indicates it's only for `WM_CLOSE`

**Review by cbracken — APPROVED** [review]

**Pull request merged**
